File: python/editdistance.py
# -*- coding: utf8 -*-
__author__ = 'arenduchintala'
import numpy as np
from optparse import OptionParser
import pdb
import traceback
from pprint import pprint


class EditDistance(object):
    def __init__(self, word2vec_file):

        self.word2vec = {}
        self.a = []
        self.b = []
        if word2vec_file is not None:
            self.loadw2v(word2vec_file)


    def edscore(self, a, b):
        ed = self.editdistance(a, b)
        edr = ed / float(max(len(a), len(b)))
        return 1.0 - edr

    def alignmentdistance(self, a, b):
        self.a = a
        self.b = b
        null_threshold = 5e-1
        alignments = {}
        for tb in b:
            align = (tb, '<eps>')
            align_max = null_threshold
            for ta in a:
                cs = self.cosine_sim(self.word2vec[tb.lower()], self.word2vec[ta.lower()])
                if cs > align_max:
                    align = (tb, ta)
                    align_max = cs
                else:
                    pass
            alignments[align] = align_max
        return alignments.keys()


    def editdistance(self, a, b):
        self.a = a
        self.b = b
        substitution_penalty = 5e-1
        inset_penalty = 5e-1
        delete_penalty = 5e-1
        table = np.zeros((len(a) + 1, len(b) + 1), dtype=float)
        came_from = {}
        # table = np.ones((len(a) + 1, len(b) + 1))
        for i in range(len(a) + 1):
            table[i, 0] = delete_penalty ** i  # i
            came_from[i, 0] = (i - 1, 0), (a[i - 1], '<eps>')

        for j in range(len(b) + 1):
            table[0, j] = inset_penalty ** j  # j
            came_from[0, j] = (0, j - 1), ('<eps>', b[j - 1])

        # print 'start'
        for i in range(1, len(a) + 1):
            for j in range(1, len(b) + 1):
                if a[i - 1] == b[j - 1]:
                    diag = table[i - 1, j - 1] * 1.0
                else:
                    if a[i - 1].lower() in self.word2vec and b[j - 1].lower() in self.word2vec:
                        cs = self.cosine_sim(self.word2vec[a[i - 1].lower()], self.word2vec[b[j - 1].lower()])
                        # print cs, a[i - 1], b[j - 1]
                    else:
                        cs = substitution_penalty
                    diag = table[i - 1, j - 1] * cs  # substitution cost
                top = table[i, j - 1] * inset_penalty  # insertion cost
                left = table[i - 1, j] * delete_penalty  # deletion cost

                best, prev, tok = max((diag, (i - 1, j - 1), (a[i - 1], b[j - 1])),
                                      (top, (i, j - 1), ('<eps>', b[j - 1])),
                                      (left, (i - 1, j), (a[i - 1], '<eps>')))

                table[i, j] = best
                came_from[i, j] = (prev, tok)
                # print 'current cell', table[i, j]
        # print table

        alignments = self.bt(came_from)
        return table[i, j], alignments

    def bt(self, cf):
        i = len(self.a)
        j = len(self.b)
        alignments = []
        while not (i == 0 and j == 0):
            prev, tok = cf[i, j]
            alignments.append((tok[0], tok[1]))  # this is an alignment
            i = prev[0]
            j = prev[1]
        alignments.reverse()
        return alignments


    def editdistance_prob(self, a, b):
        insertion_cost = np.log(0.33)
        deletion_cost = np.log(0.33)
        substitution_cost = np.log(0.34)
        table = np.zeros((len(a) + 1, len(b) + 1))
        # table = np.ones((len(a) + 1, len(b) + 1))
        for i in range(len(a) + 1):
            table[i, 0] = insertion_cost * i  # i
        for j in range(len(b) + 1):
            table[0, j] = deletion_cost * j  # j
        # print 'start'
        for i in range(1, len(a) + 1):
            for j in range(1, len(b) + 1):
                if a[i - 1] == b[j - 1]:
                    table[i, j] = table[i - 1, j - 1]
                else:
                    # print i, j
                    diag = table[i - 1, j - 1] + substitution_cost  # substitution cost
                    # print 'diag', diag
                    left = table[i - 1, j] + deletion_cost  # deletion cost
                    # print 'left', left
                    top = table[i, j - 1] + insertion_cost  # insertion cost
                    # print 'top', top
                    # best = min(diag, top, left)
                    best = max(diag, top, left)

                    # print 'best so far', best, diag, top, left
                    table[i, j] = best
                    # print 'current cell', table[i, j]
        # print table
        return table[i, j]

    def cosine_sim(self, v1, v2):
        dot = 0.0
        v1_sq = 0.0
        v2_sq = 0.0
        for i in xrange(len(v1)):
            dot += v1[i] * v2[i]
            v1_sq += v1[i] ** 2.0
            v2_sq += v2[i] ** 2.0
        denom = (np.sqrt(v1_sq) * np.sqrt(v2_sq))
        i = dot / denom
        i = i if i <= 1 else 1.0
        i = i if i >= 0.0 else 0.0
        return i


    def loadw2v(self, word2vec_file):
        try:
            open(word2vec_file, 'r')
        except IOError, err:
            print Exception, err
            return False
        print 'loading word vecs...'
        with open(word2vec_file) as infile:
            for line in infile:
                terms = line.split()
                word = terms.pop(0)
                vec = np.array(terms, dtype='|S')
                vec = vec.astype(np.float32)
                self.word2vec[word] = vec
        print 'done.'
        return True


if __name__ == "__main__":
    opt = OptionParser()
    opt.add_option("-x", dest="x", default="this is a test",
                   help="1st string")
    opt.add_option("-y", dest="y", default="this is a test",
                   help="2st string)")
    opt.add_option("-d", dest="word2vecfile", default="data/glove.6B.50d.txt",
                   help="txt file with word2vec")
    (options, _) = opt.parse_args()
    x = options.x.split()  # ", account flows down".split()  #
    # x = "A L T R U I S M".split()
    y = options.y.split()  # "it rain rains down".split()  #
    # y = "P L A S M A".split()
    E = EditDistance(options.word2vecfile)
    ed, alignments = E.editdistance(x, y)
    print ed
    print alignments
    edp = E.editdistance_prob(x, y)
    print 'ed', ed, ' ed_prob', np.exp(edp)
    # edr = edscore(x, y)
    # print 'final edr', edr

File: js/editdistance.js
/**
 * Created by arenduchintala on 8/2/15.
 */


function makeArrayOf(value, length) {
    var arr = [], i = length;
    while (i --) {
        arr[i] = value;
    }
    return arr;
}


function Create2DArray(rows, columns) {
    var arr = [];

    for (var i = 0; i < rows; i ++) {
        arr[i] = makeArrayOf(0, columns);
    }

    return arr;
}

function getTuple(a, b) {
    return [a, b];
}

function edObj(score, prev, tok) {
    var self = this
    this.score = score
    this.prev = prev
    this.tok = tok

}

function minObj(a, b, c) {
    if (a.score <= b.score) {
        if (a.score <= c.score) {
            return a
        } else {
            return c
        }
    } else {
        if (b.score <= c.score) {
            return b
        } else {
            return c
        }
    }
}

function bt(cf, arr1, arr2) {
    var i = arr1.length
    var j = arr2.length
    var alignments = []
    while (i > 0 && j > 0) {
        var edobj = cf[getTuple(i, j)]
        i = edobj.prev[0]
        j = edobj.prev[1]
        alignments.unshift(getTuple(edobj.tok[0], edobj.tok[1]))

    }
    return alignments
}

function editdistance(arr1, arr2) {
    var substitution_cost = 1
    var insertion_cost = 1
    var deletion_cost = 1
    var table = Create2DArray(arr1.length + 1, arr2.length + 1)
    var came_from = {}
    for (var i = 0; i < arr1.length + 1; i ++) {
        table[i][0] = deletion_cost * i
        came_from[getTuple(i, 0)] = getTuple(getTuple(i - 1, 0), getTuple(arr1[i - 1], '<eps>'))
    }

    for (var j = 0; j < arr2.length + 1; j ++) {
        table[0][j] = insertion_cost * j
        came_from[getTuple(0, j)] = getTuple(getTuple(0, j - 1), getTuple('<eps>', arr2[j - 1]))
    }

    var diag = null
    var top = null
    var left = null
    for (var i = 1; i < arr1.length + 1; i ++) {
        for (var j = 1; j < arr2.length + 1; j ++) {
            if (arr1[i - 1] == arr2[j - 1]) {
                diag = table[i - 1][i - 1] + 0
            } else {
                diag = table[i - 1][j - 1] + substitution_cost
            }
            top = table[i][j - 1] + insertion_cost
            left = table[i - 1][j] + deletion_cost

            var diag_obj = new edObj(diag, getTuple(i - 1, j - 1), getTuple(arr1[i - 1], arr2[j - 1]))
            var top_obj = new edObj(top, getTuple(i, j - 1), getTuple('<eps>', arr2[j - 1]))
            var left_obj = new edObj(left, getTuple(i - 1, j), getTuple(arr1[i - 1], '<eps>'))
            var best_obj = minObj(diag_obj, top_obj, left_obj)
            table[i][j] = best_obj.score
            came_from[getTuple(i, j)] = best_obj


        }
    }
    var alignments = bt(came_from, arr1, arr2)
    console.log(alignments)
    var o = {}
    o.ed = table[arr1.length][arr2.length]
    o.alignments = alignments
    return o
}

